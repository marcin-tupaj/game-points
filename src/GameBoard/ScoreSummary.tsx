import Button from 'styles/components/Button';
import { GridContainer } from 'styles/components/Grid';
import Heading from 'styles/components/Heading';

type Props = {
  bonusScore: number;
  totalScore: number;
  onNewGameClick: () => void;
};

const ScoreSummary = ({ bonusScore, totalScore, onNewGameClick }: Props) => {
  return (
    <>
      <GridContainer
        columns="auto auto"
        gap
        justifyContent="start"
        padding="lg"
      >
        <strong>Bonuses</strong>{' '}
        <span data-testid="bonus-score">{bonusScore}</span>
      </GridContainer>
      <div>
        <hr />
      </div>
      <GridContainer columns="auto auto" gap alignItems="center" padding="lg">
        <div>
          <Heading as="p" level={2} noMargin>
            Total
          </Heading>
          <Heading data-testid="total-score" as="p" level={3} noMargin>
            {totalScore}
          </Heading>
        </div>
        <div>
          <Button
            type="button"
            onClick={onNewGameClick}
            fullWidth
            aria-label="Start a new game"
          >
            New game
          </Button>
        </div>
      </GridContainer>
    </>
  );
};

export default ScoreSummary;
