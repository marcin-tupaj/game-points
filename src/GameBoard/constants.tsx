const letterItems = [
  { letter: 'A', points: 50, bonus: { points: 50, count: 3 } },
  { letter: 'B', points: 30, bonus: { points: 30, count: 2 } },
  { letter: 'C', points: 20, bonus: null },
  { letter: 'D', points: 15, bonus: null },
];

export type LetterItem = {
  letter: string;
  points: number;
  bonus: {
    points: number;
    count: number;
  } | null;
};

export const getLetterItems = (): LetterItem[] => letterItems;
