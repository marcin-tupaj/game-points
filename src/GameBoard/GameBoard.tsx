import { useMemo, useState } from 'react';
import { LetterItem, getLetterItems } from './constants';
import Button from 'styles/components/Button';
import {
  GameContainer,
  GridContainer,
  GameSection,
  PageContainer,
  ScoreSection,
  SectionTitle,
} from 'styles/components/Grid';
import GlobalStyle from 'styles/Global';
import ItemBox from 'styles/components/ItemBox';
import { itemColors } from 'styles/pages/HomePage/constants';
import Heading from 'styles/components/Heading';
import {
  SectionBody,
  ScoreTableColumn,
  ScoreTable,
  ScoreTableRow,
  ScoreTableCell,
} from 'styles/pages/HomePage/ScoreTable';

export default function GameBoard() {
  const [itemsState, setItemsState] = useState<{
    [key: string]: { count: number; score: number; bonusScore: number };
  }>({});

  const letterItems = getLetterItems();

  const handleLetterClick = (clikedLetter: LetterItem) => () => {
    setItemsState(prev => {
      const newCount = (prev[clikedLetter.letter]?.count || 0) + 1;
      const newScore =
        (prev[clikedLetter.letter]?.score || 0) + clikedLetter.points;
      const newBonusScore = clikedLetter.bonus
        ? Math.floor(newCount / clikedLetter.bonus.count) *
          clikedLetter.bonus.points
        : 0;
      return {
        ...prev,
        [clikedLetter.letter]: {
          count: newCount,
          score: newScore,
          bonusScore: newBonusScore,
        },
      };
    });
  };

  const overallScore = useMemo(
    () =>
      Object.keys(itemsState).reduce((acc, letter) => {
        const letterItem = itemsState[letter];
        return acc + letterItem.score;
      }, 0),
    [itemsState]
  );

  const bonusScore = useMemo(
    () =>
      Object.keys(itemsState).reduce((acc, letter) => {
        const letterItem = itemsState[letter];
        return acc + letterItem.bonusScore;
      }, 0),
    [itemsState]
  );

  const handleNewGame = () => {
    setItemsState({});
  };

  return (
    <>
      <GlobalStyle />
      <PageContainer>
        <GameContainer columns="2fr 1fr">
          <GameSection>
            <SectionTitle>Kahoot! points</SectionTitle>
            <SectionBody>
              <GridContainer columns="2, 1fr" repeat gap>
                {letterItems.map(
                  (letterItem, index) =>
                    letterItem && (
                      <Button
                        type="button"
                        key={letterItem.letter}
                        variant={index % itemColors.length}
                        onClick={handleLetterClick(letterItem)}
                        aria-label={`Select letter ${letterItem.letter}`}
                      >
                        {letterItem.letter}
                      </Button>
                    )
                )}
              </GridContainer>
            </SectionBody>
          </GameSection>
          <ScoreSection>
            <SectionTitle id="scores-section-title">Player items</SectionTitle>
            <ScoreTableColumn>
              <ScoreTable aria-describedby="scores-section-title">
                <thead>
                  <ScoreTableRow>
                    <ScoreTableCell as="th" scope="col">
                      Item
                    </ScoreTableCell>
                    <ScoreTableCell as="th" scope="col">
                      Quantity
                    </ScoreTableCell>
                    <ScoreTableCell as="th" scope="col">
                      Score
                    </ScoreTableCell>
                  </ScoreTableRow>
                </thead>
                <tbody>
                  {!Object.keys(itemsState).length ? (
                    <ScoreTableRow>
                      <ScoreTableCell colSpan={3}>
                        Choose your item!
                      </ScoreTableCell>
                    </ScoreTableRow>
                  ) : (
                    Object.keys(itemsState).map(counterKey => (
                      <ScoreTableRow key={counterKey}>
                        <ScoreTableCell>
                          <ItemBox
                            variant={
                              letterItems.findIndex(
                                item => item.letter === counterKey
                              ) % itemColors.length
                            }
                            aria-label={`Item ${counterKey}`}
                          >
                            <span>{counterKey}</span>
                          </ItemBox>
                        </ScoreTableCell>
                        <ScoreTableCell
                          aria-label={`Quantity: ${itemsState[counterKey].count}`}
                        >
                          {itemsState[counterKey].count}
                        </ScoreTableCell>
                        <ScoreTableCell
                          aria-label={`Score: ${
                            itemsState[counterKey].score +
                            itemsState[counterKey].bonusScore
                          }`}
                        >
                          {itemsState[counterKey].score +
                            itemsState[counterKey].bonusScore}
                        </ScoreTableCell>
                      </ScoreTableRow>
                    ))
                  )}
                </tbody>
              </ScoreTable>
            </ScoreTableColumn>
            <GridContainer
              columns="auto auto"
              gap
              justifyContent="start"
              padding="lg"
            >
              <strong>Bonuses</strong>{' '}
              <span data-testid="bonus-score">{bonusScore}</span>
            </GridContainer>
            <div>
              <hr />
            </div>
            <GridContainer
              columns="auto auto"
              gap
              alignItems="center"
              padding="lg"
            >
              <div>
                <Heading as="p" level={2} noMargin>
                  Total
                </Heading>
                <Heading data-testid="total-score" as="p" level={3} noMargin>
                  {overallScore + bonusScore}
                </Heading>
              </div>
              <div>
                <Button
                  type="button"
                  onClick={handleNewGame}
                  fullWidth
                  aria-label="Start a new game"
                >
                  New game
                </Button>
              </div>
            </GridContainer>
          </ScoreSection>
        </GameContainer>
      </PageContainer>
    </>
  );
}
