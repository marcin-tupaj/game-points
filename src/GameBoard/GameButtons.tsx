import Button from 'styles/components/Button';
import { GridContainer } from 'styles/components/Grid';
import { itemColors } from 'styles/pages/HomePage/constants';
import { ItemsState, LetterItem } from 'types';
import { Dispatch, SetStateAction, memo } from 'react';

type Props = {
  letterItems: LetterItem[];
  setItemsState: Dispatch<SetStateAction<ItemsState>>;
};

const GameButtons = memo(({ letterItems, setItemsState }: Props) => {
  const handleLetterClick = (clikedLetter: LetterItem) => () => {
    setItemsState(prev => {
      const newCount = (prev[clikedLetter.letter]?.count || 0) + 1;
      const newScore =
        (prev[clikedLetter.letter]?.score || 0) + clikedLetter.points;
      const newBonusScore = clikedLetter.bonus
        ? Math.floor(newCount / clikedLetter.bonus.count) *
          clikedLetter.bonus.points
        : 0;

      return {
        ...prev,
        [clikedLetter.letter]: {
          count: newCount,
          score: newScore,
          bonusScore: newBonusScore,
        },
      };
    });
  };

  return (
    <GridContainer columns="2, 1fr" repeat gap>
      {letterItems.map(
        (letterItem, index) =>
          letterItem && (
            <Button
              type="button"
              key={letterItem.letter}
              variant={index % itemColors.length}
              onClick={handleLetterClick(letterItem)}
              aria-label={`Select letter ${letterItem.letter}`}
            >
              {letterItem.letter}
            </Button>
          )
      )}
    </GridContainer>
  );
});

export default GameButtons;
