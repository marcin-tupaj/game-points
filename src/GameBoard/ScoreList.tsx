import { ItemsState, LetterItem } from 'types';
import ItemBox from 'styles/components/ItemBox';
import {
  ScoreTable,
  ScoreTableCell,
  ScoreTableRow,
} from 'styles/pages/HomePage/ScoreTable';
import { itemColors } from 'styles/pages/HomePage/constants';

type Props = {
  itemsState: ItemsState;
  letterItems: LetterItem[];
};

const ScoreList = ({ itemsState, letterItems }: Props) => {
  return (
    <ScoreTable aria-describedby="scores-section-title">
      <thead>
        <ScoreTableRow>
          <ScoreTableCell as="th" scope="col">
            Item
          </ScoreTableCell>
          <ScoreTableCell as="th" scope="col">
            Quantity
          </ScoreTableCell>
          <ScoreTableCell as="th" scope="col">
            Score
          </ScoreTableCell>
        </ScoreTableRow>
      </thead>
      <tbody>
        {!Object.keys(itemsState).length ? (
          <ScoreTableRow>
            <ScoreTableCell colSpan={3}>Choose your item!</ScoreTableCell>
          </ScoreTableRow>
        ) : (
          Object.keys(itemsState).map(counterKey => (
            <ScoreTableRow key={counterKey}>
              <ScoreTableCell>
                <ItemBox
                  variant={
                    letterItems.findIndex(item => item.letter === counterKey) %
                    itemColors.length
                  }
                  aria-label={`Item ${counterKey}`}
                >
                  <span>{counterKey}</span>
                </ItemBox>
              </ScoreTableCell>
              <ScoreTableCell
                aria-label={`Quantity: ${itemsState[counterKey].count}`}
              >
                {itemsState[counterKey].count}
              </ScoreTableCell>
              <ScoreTableCell
                aria-label={`Score: ${
                  itemsState[counterKey].score +
                  itemsState[counterKey].bonusScore
                }`}
              >
                {itemsState[counterKey].score +
                  itemsState[counterKey].bonusScore}
              </ScoreTableCell>
            </ScoreTableRow>
          ))
        )}
      </tbody>
    </ScoreTable>
  );
};
export default ScoreList;
