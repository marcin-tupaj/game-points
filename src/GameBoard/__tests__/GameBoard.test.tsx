import { render } from '@testing-library/react';
import GameBoard from '../GameBoard';
import '@testing-library/jest-dom';

test('renders the expected text', () => {
  const { getByText } = render(<GameBoard />);

  const textElement = getByText('Game Board');

  expect(textElement).toBeInTheDocument();
});
