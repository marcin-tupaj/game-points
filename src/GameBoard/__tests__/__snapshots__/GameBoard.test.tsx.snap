// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`GameBoard should render the correct bonus and total scores after clicking multiple letter buttons 1`] = `
<div
  class="Grid__PageContainer-sc-dhuxyx-1 jqwcJ"
>
  <div
    class="Grid__GridContainer-sc-dhuxyx-0 Grid__GameContainer-sc-dhuxyx-2 ioKryS MxoDc"
  >
    <main
      class="Grid__GameSection-sc-dhuxyx-3"
    >
      <h1
        class="Heading-sc-na8skm-0 Grid__SectionTitle-sc-dhuxyx-5 ijWcIm csBosv"
      >
        Kahoot! points
      </h1>
      <div
        class="ScoreTable__SectionBody-sc-1gj21se-2 kNjoI"
      >
        <div
          class="Grid__GridContainer-sc-dhuxyx-0 OPYyZ"
        >
          <button
            aria-label="Select letter A"
            class="Button-sc-18nym9o-0 grkJOj"
            type="button"
          >
            A
          </button>
          <button
            aria-label="Select letter B"
            class="Button-sc-18nym9o-0 juqerL"
            type="button"
          >
            B
          </button>
          <button
            aria-label="Select letter C"
            class="Button-sc-18nym9o-0 ddpfln"
            type="button"
          >
            C
          </button>
          <button
            aria-label="Select letter D"
            class="Button-sc-18nym9o-0 YlkwH"
            type="button"
          >
            D
          </button>
        </div>
      </div>
    </main>
    <aside
      class="Grid__ScoreSection-sc-dhuxyx-4 dshiz"
    >
      <h1
        class="Heading-sc-na8skm-0 Grid__SectionTitle-sc-dhuxyx-5 ijWcIm csBosv"
        id="scores-section-title"
      >
        Player items
      </h1>
      <div
        class="ScoreTable__ScoreTableColumn-sc-1gj21se-0 kQqoOu"
      >
        <table
          aria-describedby="scores-section-title"
          class="ScoreTable-sc-1gj21se-3 iNdZGG"
        >
          <thead>
            <tr
              class="ScoreTable__ScoreTableRow-sc-1gj21se-4 iOnbct"
            >
              <th
                class="ScoreTable__ScoreTableCell-sc-1gj21se-5 TligK"
                scope="col"
              >
                Item
              </th>
              <th
                class="ScoreTable__ScoreTableCell-sc-1gj21se-5 TligK"
                scope="col"
              >
                Quantity
              </th>
              <th
                class="ScoreTable__ScoreTableCell-sc-1gj21se-5 TligK"
                scope="col"
              >
                Score
              </th>
            </tr>
          </thead>
          <tbody>
            <tr
              class="ScoreTable__ScoreTableRow-sc-1gj21se-4 iOnbct"
            >
              <td
                class="ScoreTable__ScoreTableCell-sc-1gj21se-5 TligK"
              >
                <div
                  aria-label="Item A"
                  class="ItemBox-sc-l5pjw0-0 gltgDI"
                >
                  <span>
                    A
                  </span>
                </div>
              </td>
              <td
                aria-label="Quantity: 4"
                class="ScoreTable__ScoreTableCell-sc-1gj21se-5 TligK"
              >
                4
              </td>
              <td
                aria-label="Score: 600"
                class="ScoreTable__ScoreTableCell-sc-1gj21se-5 TligK"
              >
                600
              </td>
            </tr>
            <tr
              class="ScoreTable__ScoreTableRow-sc-1gj21se-4 iOnbct"
            >
              <td
                class="ScoreTable__ScoreTableCell-sc-1gj21se-5 TligK"
              >
                <div
                  aria-label="Item B"
                  class="ItemBox-sc-l5pjw0-0 iXxPJA"
                >
                  <span>
                    B
                  </span>
                </div>
              </td>
              <td
                aria-label="Quantity: 1"
                class="ScoreTable__ScoreTableCell-sc-1gj21se-5 TligK"
              >
                1
              </td>
              <td
                aria-label="Score: 80"
                class="ScoreTable__ScoreTableCell-sc-1gj21se-5 TligK"
              >
                80
              </td>
            </tr>
            <tr
              class="ScoreTable__ScoreTableRow-sc-1gj21se-4 iOnbct"
            >
              <td
                class="ScoreTable__ScoreTableCell-sc-1gj21se-5 TligK"
              >
                <div
                  aria-label="Item C"
                  class="ItemBox-sc-l5pjw0-0 jxgbKM"
                >
                  <span>
                    C
                  </span>
                </div>
              </td>
              <td
                aria-label="Quantity: 1"
                class="ScoreTable__ScoreTableCell-sc-1gj21se-5 TligK"
              >
                1
              </td>
              <td
                aria-label="Score: 60"
                class="ScoreTable__ScoreTableCell-sc-1gj21se-5 TligK"
              >
                60
              </td>
            </tr>
            <tr
              class="ScoreTable__ScoreTableRow-sc-1gj21se-4 iOnbct"
            >
              <td
                class="ScoreTable__ScoreTableCell-sc-1gj21se-5 TligK"
              >
                <div
                  aria-label="Item D"
                  class="ItemBox-sc-l5pjw0-0 kDQgoU"
                >
                  <span>
                    D
                  </span>
                </div>
              </td>
              <td
                aria-label="Quantity: 1"
                class="ScoreTable__ScoreTableCell-sc-1gj21se-5 TligK"
              >
                1
              </td>
              <td
                aria-label="Score: 40"
                class="ScoreTable__ScoreTableCell-sc-1gj21se-5 TligK"
              >
                40
              </td>
            </tr>
          </tbody>
        </table>
      </div>
      <div
        class="Grid__GridContainer-sc-dhuxyx-0 eESGYv"
      >
        <strong>
          Bonuses
        </strong>
         
        <span
          data-testid="bonus-score"
        >
          200
        </span>
      </div>
      <div>
        <hr />
      </div>
      <div
        class="Grid__GridContainer-sc-dhuxyx-0 fNbZqM"
      >
        <div>
          <p
            class="Heading-sc-na8skm-0 kNUXZO"
          >
            Total
          </p>
          <p
            class="Heading-sc-na8skm-0 bUvPQD"
            data-testid="total-score"
          >
            780
          </p>
        </div>
        <div>
          <button
            aria-label="Start a new game"
            class="Button-sc-18nym9o-0 lbBiOI"
            type="button"
          >
            New game
          </button>
        </div>
      </div>
    </aside>
  </div>
</div>
`;
